packages/adblocker-electron/src/index.ts

```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
   // ElectronBlocker-specific additions to FiltersEngine
   // ----------------------------------------------------------------------- //
 
-  public onIsMutationObserverEnabled = (event: Electron.IpcMainEvent): void => {
-    event.returnValue = this.config.enableMutationObserver;
+  public onIsMutationObserverEnabled = async (_: Electron.IpcMainInvokeEvent): Promise<boolean> => {
+    return this.config.enableMutationObserver;
   };
 
-  public onGetCosmeticFiltersFirst = (event: Electron.IpcMainEvent, url: string) => {
-    // Extract hostname from sender's URL
+  public onInjectCosmeticFilters = async (
+    event: Electron.IpcMainInvokeEvent,
+    url: string,
+    msg?: IBackgroundCallback,
+  ): Promise<{ active: boolean; error?: string }> => {
     const parsed = parse(url);
     const hostname = parsed.hostname || '';
     const domain = parsed.domain || '';
 
+    // `msg` is undefined for the initial call and present for subsequent updates
     const { active, styles, scripts, extended } = this.getCosmeticsFilters({
       domain,
       hostname,
       url,
 
-      // This needs to be done only once per frame
-      getBaseRules: true,
-      getInjectionRules: true,
-      getExtendedRules: true,
-      getRulesFromHostname: true,
-      getRulesFromDOM: false, // Only done on updates (see `onGetCosmeticFiltersUpdated`)
+      // DOM information, only available for updates
+      classes: msg?.classes,
+      hrefs: msg?.hrefs,
+      ids: msg?.ids,
+
+      // Rules to fetch: true for initial call, false for updates
+      getBaseRules: !msg,
+      getInjectionRules: !msg,
+      getExtendedRules: !msg,
+      getRulesFromHostname: !msg,
+
+      // Only true for update calls when we have DOM information
+      getRulesFromDOM: !!msg,
 
       callerContext: {
         frameId: event.frameId,
         processId: event.processId,
+        lifecycle: msg?.lifecycle,
```

Comment:
If you have a nice use-case to this addition to the caller context, please let us know. It'll help our future direction in the implementation. The caller context declarations you can see in many places were added to help debugging.

Reply:
Nothing off the top of my head right now, but I'll keep this in mind.